File: 0113-path-sum-ii/0113-path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>>ans;
    void dfs(TreeNode* root, int targetSum, vector<int>temp) {
        if(!root) return;
        temp.push_back(root->val);
        if(!root->left && !root->right && root->val == targetSum) {
            ans.push_back(temp);
        }
        dfs(root->left, targetSum - root->val, temp);
        dfs(root->right, targetSum - root->val, temp);
        temp.pop_back();
    }
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<int>temp;
        dfs(root, targetSum, temp);
        return ans;
    }
};

File: 0113-path-sum-ii/0113-path-sum-ii.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    List<List<Integer>>ans = new ArrayList<>();
    void solve(TreeNode root, int target, List<Integer>temp) {
        if(root == null) return;
        temp.add(root.val);
        if(root.left == null && root.right == null && root.val == target) {
            ans.add(new ArrayList<>(temp));
        }
        solve(root.left, target - root.val, temp);
        solve(root.right, target - root.val, temp);
        temp.remove(temp.size() - 1);
    }
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<Integer>temp = new ArrayList<>();
        solve(root, targetSum, temp);
        return ans;
    }
}